app/controllers/application_controller.rb

```diff
@@ -1,5 +1,31 @@
 class ApplicationController < ActionController::Base
+  ALLOWED_ORIGINS = ENV.fetch('ALLOWED_ORIGINS').split(',').freeze
   # Prevent CSRF attacks by raising an exception.
   # For APIs, you may want to use :null_session instead.
   # protect_from_forgery with: :exception
+
+  after_action :cors_set_access_control_headers
+
+  def cors_set_access_control_headers
+    http_origin = request.headers['HTTP_ORIGIN']
+
+    if http_origin
+      headers['Access-Control-Allow-Origin'] = http_origin
+      headers['Access-Control-Allow-Methods'] = 'POST, GET, PUT, DELETE, OPTIONS'
+      headers['Access-Control-Allow-Credentials'] = 'true'
+    end
+  end
+
+  def cors_preflight_check
+    if request.method == 'OPTIONS'
+      if ALLOWED_ORIGINS.include?(request.headers['HTTP_ORIGIN'])
+        headers['Access-Control-Allow-Origin'] = request.headers['HTTP_ORIGIN']
+        headers['Access-Control-Allow-Credentials'] = 'true'
+        headers['Access-Control-Allow-Methods'] = 'POST, GET, PUT, DELETE, OPTIONS'
+        headers['Access-Control-Allow-Headers'] = 'accept,content-type'
+      end
+
+      render :text => '', :content_type => 'text/plain'
```

Comment:
Pretty sure Rack/Rails should handle that (although never hurts to double-check). Also I think `render :text` implies `Content-Type: text/plain`.